### DISCUSSION TITLE
Hopper sensor issue

### DISCUSSION THREAD

**Discussion by thomasstrach** [nebhead/PiFire#83] [Q&A]
I am using the VL53L0X Tof sensor and the display always stays with 100% hopper level.
So I tried a second one with the same outcome. 
Originally, last year,  I had the problem that the value was altering (protype SW version) but with the new releases I end up with this problem. Any idea how to solve or investigate in this?

**Comment by weberbox** [comment]
Did you select the VL53L0X Tof  when you setup PiFire? When installing it defaults to prototype which will always show 100% regardless if hopper sensor is attached or not.

You can change this setting with the modules.sh script. 

ssh into your pi
```
cd ~/pifire
bash modules.sh
```
then you should probably reboot the pi when finished. 

```
sudo reboot now
```

**Reply by thomasstrach** [reply]
Yes, I did this setup. 
It is most unlikely that 2 new Tof sensors are not working.
What I did additionally is this setting, might it have some impact?
![image](https://user-images.githubusercontent.com/50183687/153560264-ca461afd-6d69-45b1-9a0d-b33961817540.png)

**Reply by nebhead** [reply]
Try backing out those changes to do faster baud rate and doing a reboot.  I don't really think they are necessary, as the architecture has changed enough that reading the temperatures from I2C seems to work just fine now.  In older versions, I was seeing some values drop out from time to time - which I assumed was a speed issue.  But I don't think it was.

**Reply by thomasstrach** [reply]
I hashtags the changes and still do not get any value from the sensors. Is there any logfile I can check?

**Comment by nebhead** [comment]
Yes, let's check your control.err.log and control.out.log in the
~/pifire/logs directory.  Can you post the output of those?

Also, worth checking your settings.json (just the modules section) to make
sure it's setup correctly.  Should look something like this:

```
"modules": {
"adc": "prototype",
"display": "prototype",
"dist": "prototype",
"grillplat": "prototype"
},
```

I would expect none of these to be prototype on yours.

**Comment by thomasstrach** [comment]
Find attached the requested 2 files.
In the settings.json file I have these values: 
"modules": {
    "adc": "ads1115",
    "display": "ili9341b",
    "dist": "vl53l0x",
    "grillplat": "pifire"
  },

[control.out.log](https://github.com/nebhead/PiFire/files/8068434/control.out.log)
[control.err.log](https://github.com/nebhead/PiFire/files/8068435/control.err.log)

Hope you can find some more details.

**Reply by nebhead** [reply]
Thanks!  OK, so your settings are correct, so we can rule out any issues with that part.  

As for the log files, I'm seeing a lot of SPI errors on the ADC side in the 'control.err.log' file.  Those might be old errors, however, it's something to look at.  Are your temperature readings working correctly right now?   It's possible that these errors happened previously if there wasn't an ADS1115 plugged in, or the I2C wasn't enabled.  

As for the 'control.out.log', we can see that the range finder is having some trouble talking to the I2C device via the API.  

This is what it should look like:
```
VL53L0X_GetDeviceInfo:
Device Name : VL53L0X ES1 or later
Device Type : VL53L0X
Device ID : VL53L0CBV0DH/1$1
ProductRevisionMajor : 1
ProductRevisionMinor : 1
API Status: 0 : No Error
VL53L0X_BETTER_ACCURACY_MODE
API Status: 0 : No Error 
```

But your log has an error here:

```
Call of VL53L0X_PerformRefCalibration
API Status: -20 : Control Interface Error
VL53L0X_BETTER_ACCURACY_MODE
Set Accuracy
API Status: -20 : Control Interface Error
VL53L0X Start Ranging Address 0x29
```

The module can't seem to communicate with your device.  This may mean you have a physical issue with the I2C bus.  I would recommend re-confirming that your I2C pins and wiring is correct.  If you are using the PCB, make sure the solder joints to the I2C bus pins are flowed well.  Maybe reflow them.

**Reply by thomasstrach** [reply]
I have checked the connections on the PCB and there was no issue at all. So I decided to buy a different Tof sensor and with this 3rd new sensor all,is working fine.
Here is the link to the 2 ones which did not work at all

https://www.az-delivery.de/en/products/vl53l0x-time-of-flight-tof-laser-abstandssensor

Maybe you have an idea why these are not working? They look like many others on Amazon.

This one here is working fine: 

https://www.amazon.de/dp/B07VJ87PZC/ref=cm_sw_em_r_mt_dp_FVEQ6TTJZK8QERQKB75E?_encoding=UTF8&psc=1

Thanks a lot for having a look into my files and supporting me with my topic.
I really love the whole project and the upcoming developments look,promising as well.

**Comment by FastEddie49** [comment]
I am  just finishing my build and I ran into a problem with the VL53L0X Tof sensor...it would work correctly with a short cable but not with a cable that is about 1m long. I was using 28AWG ribbon cable with parallel wires and I switched to some stranded 24AWG twisted pair that I stripped out of some Cat6 UTP ethernet cable. Problem solved. I think the long length of parallel wires was adding capacitance and reflectance along the cable that was causing interference with the sensor. Something to check if you are having issues. Prior to changing the wiring I wasn't getting an address for the sensor on the I2C bus (0x29). Adafruit has a device list, BTW https://learn.adafruit.com/i2c-addresses/the-list

![image](https://user-images.githubusercontent.com/42302231/197374126-d8121d15-833e-483a-8333-38675773b53e.png)

**Reply by FastEddie49** [reply]
This turned out to be a less than perfect solution. After installing the ToF on the hopper, I'm getting inconsistent reporting from the sensor. Originally, even with soldering the ribbon cable directly to the headers, the capacitance/resistance of the ribbon cable interfered with getting an I2C address from the sensor. Now I'm getting a consistent address, but the distance reporting is spotty, sometimes timing out or reporting 0% (if the sensor isn't present, 100% is reported). I've ordered an inexpensive Adafruit LTC4311 I2C Extender/Active Terminator that will hopefully clean up the signal from the ToF sensor.

**Reply by nebhead** [reply]
One other thing to try, is adding 10k pullups (to 3.3V) on the SCA/SDA lines, near the TOF sensor, if you haven't tried this already.

**Reply by FastEddie49** [reply]
I suppose I could solder up something on a female header and plug it in on the bus to try it out...

**Reply by FastEddie49** [reply]
And I'm kicking myself now but I should have used the gold-plated pins on those JST connectors that I forgot I had and just discovered this morning!

**Reply by FastEddie49** [reply]
Re-reading your reply, "_near the TOF sensor_"...putting the pullups on the bus likely won't help. Not one of my better ideas.
